### DISCUSSION TITLE
ST7789 (V2)

### DISCUSSION THREAD

**Discussion by richcj10** [Bodmer/TFT_eSPI#1793] [Q&A - General]
I have a display that states its a ST7789V2. It works well with the library from Adafruit. But has color problems with TFT_eSPI as well as the the display image is shifted up. (with unused pixels at the bottom) 

I have been looking at making modifications to the x/y size and position to the display. But making changes in Rotation.h (ST7789_2) doesn't seem to effect the outcome. Something else seems to be overwriting the changes. Any other locations I need to look at?

**Comment by Bodmer** [comment]
#define CGRAM_OFFSET
Needs to be defined somewhere for the offsets to be used. [See here](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_Drivers/ST7789_Defines.h) for known display variants.

Example rotaion offsets are [invoked here](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_Drivers/ST7789_Rotation.h).

Not sure if all these have been copied to the _2 code variant.

You may need to call setRotation after tft.init() so the offsets get applied.

**Reply by richcj10** [reply]
That is what I am doing now. seems to keep with the same issue. I have a logic analyzer. I am going to probe the data line and see where the depart in config happens.

**Comment by Bodmer** [comment]
The colour problems are typically red and blue swap or inversion (blak = white and vice versa). There are setup options to correct these as they vary between suppliers.

Ofsets are determined in setRotation and applied for each graphics operation, [see here](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.cpp#L3212-L3219).

What pixel pixel count w and h does your display have?  Do you know what offsets are needed?

Post a copy of you setup as a zip file, there may be options you have not noticed.

There is a setViewport() command that can move the display origin and define a window size anywhere in the native display area so this is an alternative approach to dealing with offsets.

The variables [colstart etc are private ](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.h#L787-L790)so cannot be accessed from a sketch. But they could be moved to the public zone and then accessed from a sketch. ie  Serial.println(tft.colstart);

**Reply by richcj10** [reply]
280x240.
[Setup71_ESP32_S2_ST7789.zip](https://github.com/Bodmer/TFT_eSPI/files/8551472/Setup71_ESP32_S2_ST7789.zip)

**Comment by Bodmer** [comment]
For non-native widths and heights it is expected that this is defined in the setup, otherwise defaults may be used (240x320). So adding this will prevent over-ride:
#define TFT_WIDTH  240
#define TFT_HEIGHT 280

**Reply by richcj10** [reply]
Where should that be placed? I tried that in user configuration. Didn't move it.

**Reply by Bodmer** [reply]
In the setup file with pin definitions etc.

**Reply by richcj10** [reply]
[Setup71_ESP32_S2_ST7789.zip](https://github.com/Bodmer/TFT_eSPI/files/8559873/Setup71_ESP32_S2_ST7789.zip)
This is the video showing what the above config does. https://youtube.com/shorts/ITtnvjG3ok4

The display w/ Adafruit code looks just like this: https://www.adafruit.com/product/5206

**Reply by Bodmer** [reply]
I have added [these lines](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_Drivers/ST7789_Defines.h#L22-L29) to the init file for that display driver. This should make the offsets work.

It appears the display is inverted too.  It is not clear if red and blue are swapped too. Add uncommented defines to your setup as appropriate to correct this:

```
// #define TFT_INVERSION_ON
// #define TFT_INVERSION_OFF

//  #define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//  #define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

**Reply by richcj10** [reply]
Colors, Fixed :) 

Display position, not yet. :( Same as before. 

I hope to use the logic analyzer on the display tomorrow. I had to leave it at work.

**Comment by richcj10** [comment]
Changing the init line to this:

TFT_eSPI tft = TFT_eSPI(300,240);  

Gave me assess to the entire display. Tho there still is an offset issue. 

  tft.setCursor(30, 20, 2);
  // // large block of text
  tft.setTextColor(TFT_WHITE,TFT_BLACK);  tft.setTextSize(1);
  // We can now plot text on screen using the "print" class
  tft.println("Hello World!");

Puts the H right at the Left most of the display. (Instead of 0,20)

Changing values in ST7789_Rotation.h doesn't affect the displays position.

**Reply by Bodmer** [reply]
The offsets can be corrected using setViewport, that worked for another dsplay with offsets here:
https://github.com/Bodmer/TFT_eSPI/issues/1800#issuecomment-1111511650

**Reply by Bodmer** [reply]
Incidentally sizes must always be in the portrait orientation, so this would be the correct way to specifiy the screen size:
TFT_eSPI tft = TFT_eSPI(240,300);

However there are a few displays (e.g. ILI9342) that by default start with landscape orientation so it can get confusing. The library uses the not native pixel count to detect offsets. The native resolution for the ST7789 is 240 x 320, so the library looks for the initial set height of 300 in your case.  The position of your screen within this native area needs to be determined by experiment or the vendors info.

**Reply by richcj10** [reply]
Thanks, I will Play with this tonight. Your support has been great.

**Reply by Bodmer** [reply]
I have updated the master library to add the required offsets to the ST7789_2_DRIVER. They were only in the ST7789_DRIVER files.

**Reply by richcj10** [reply]
Do you have a tip jar? AKA buy you a coffee?

**Reply by Bodmer** [reply]
No, thank you,  just give a donation to a charity of your choice instead.

**Reply by elgerg** [reply]
@Bodmer are you an angel?

**Reply by ForbiddenEra** [reply]
Nice. Also I read that waveshare (or maybe all variations) have a 20px offset. I was looking at the display datasheet and it supports hardware scrolling (at least vertically..?), so it would make sense that the frame buffer would be taller than the actual screen.
